Refactor sandbox preload initialization.

tarruda opened the pull request:
Use a single synchronous IPC call to retrieve data required by early
sandbox scripts. This has two purposes:

- Optimize preload script initialization by:
  - Using one synchronous IPC call to retrieve preload script,
  webContentsId (more on that later), process.{platform,execPath,env}
  - Lazy loading as many modules as possible.
- Fix #12316 for sandbox. @MarshallOfSound addressed the issue in
  #12342, but it was still present in sandbox mode. By loading
  webContentsId very early and skipping remote module at early
  startup, we fix it for sandbox.

miniak commented:
This PR is in conflict with mine https://github.com/electron/electron/pull/12832.

miniak commented on lib/browser/rpc-server.js:
```diff
@@ -449,3 +450,13 @@ ipcMain.on('ELECTRON_BROWSER_WINDOW_CLOSE', function (event) {
   event.returnValue = null
 })
+
+ipcMain.on('ELECTRON_BROWSER_SANDBOX_LOAD', function (event, preloadPath) {
```
I would not pass the `preloadPath` as an argument. We know which webContents sent the IPC so it should be possible to get the preload script path from the webPreferences.

miniak replied:
Apart from this issue I like this change, batching multiple synchronous IPC calls into make to make loading faster is definitely welcome 👍

miniak replied:
We are actually planning to add some filtering for `remote.require`, `remote.getGlobal`, etc.

miniak replied:
In my case I had to keep the preloadPath as renderer argument as I need to display an error in case the preload script cannot be loaded. How do you handle that case? Currently we get an error in JS console with the path to the preload script.

tarruda replied:
> We are actually planning to add some filtering for remote.require, remote.getGlobal, etc.

Cool, that would be a welcome feature :+1: 

> In my case I had to keep the preloadPath as renderer argument as I need to display an error in case the preload script cannot be loaded. How do you handle that case? Currently we get an error in JS console with the path to the preload script.

`ELECTRON_BROWSER_SANDBOX_LOAD` can return an error if it fails to load the script

miniak replied:
Returning an error would be great. We should not have a silent failure without any logs in the console.

miniak replied:
Also if you just let fs.readFileSync throw and not assign the returnValue, sendSync in the renderer process will get stuck waiting.

MarshallOfSound replied:
I agree with @miniak this is basically a one-liner "read any file plz" API, I'd rather we (in this PR) used `event.sender.getLastWebPreferences()`

miniak replied:
@tarruda this is the last remaining issue

tarruda replied:
> I agree with @miniak this is basically a one-liner "read any file plz" API, I'd rather we (in this PR) used event.sender.getLastWebPreferences()

We currently give access to `electron.remote.fs` which can not only read any file, but also invoke any FS function. There's no regression in terms of security here by allowing the renderer to specify the preloadPath.

These are my arguments for keeping the current behavior and addressing this issue in a separate PR:

- If we start using `event.sender.getLastWebPreferences()`, there's no  reason to pass the preload path as argument to the renderer command line, so it would become necessary to refactor more code in C++ both in browser/renderer. (Unless we are OK with leaving dead code to be handled in a separate PR)
- There's no guarantee that this will work as expected without introducing new regressions. Maybe it will, but I will have to investigate if `preload` will be present in every `event.sender.webPreferences` (in windows created by window.open, for example, which follow a different webContents creation path).
- This is a completely separate change, not related to the goal of this PR which is to fix  #12316. I'd rather push refactoring changes in different PRs.

If you guys still insist that this should be done in the current PR, I will do it, but not right now since it will require more testing/verification which I can't do this week.

miniak replied:
@MarshallOfSound? I guess I am ok if this gets addressed in a separate PR.

miniak commented on lib/sandboxed_renderer/init.js:
```diff
@@ -43,15 +47,16 @@ preloadProcess.hang = () => binding.hang()
 function preloadRequire (module) {
   if (preloadModules.has(module)) {
-    return preloadModules.get(module)
+    return require(module)
```
Does this work with browserify?

miniak replied:
What is confusing to me is that we have `fs.js`, `os.js`, `path.js` in `lib/sandboxed_renderer/api/exports`, which are also listed in electron.gyp, but there is no `timers.js` or `url.js`

miniak replied:
I see, browserify has bundled support for those

miniak replied:
According to the docs
> Browserify parses the AST for require() calls to traverse the entire dependency graph of your project.

doesn't that mean that all the require calls have to have the module name as literal string?

miniak commented:
@tarruda off-topic question, why is `clearImmediate` not exposed to the preload script, only `setImmediate`?

miniak commented on lib/sandboxed_renderer/init.js:
```diff
@@ -20,19 +20,23 @@ for (let prop of Object.keys(events.EventEmitter.prototype)) {
+const preloadModules = new Set([
+  'child_process',
+  'electron',
+  'fs',
```
why did you remove events?

miniak replied:
It was added recently in https://github.com/electron/electron/pull/12828

tarruda replied:
It was a merge error, I took this patch from an older electron version we use in our app. Will add it back :+1:

tarruda commented:
> @tarruda off-topic question, why is clearImmediate not exposed to the preload script, only setImmediate?

I forgot to add it when this file was created. Will add it now :+1:

tarruda commented:
@miniak about this change:

> In any case, I agree that it would be cleaner to take preloadPath from webPreferences directly as it would also remove the need to pass the path as a renderer argument.

Since it involves some C++ changes I will do it in a separate PR ok?

miniak commented on lib/sandboxed_renderer/init.js:
```diff
@@ -43,15 +51,19 @@ preloadProcess.hang = () => binding.hang()
+    return loadedModules.get(module)
+  }
+  if (remoteModules.has(module)) {
+    return require(module)
```
shouldn't this be electron.remote.require instead?

tarruda replied:
No, we still require these files to emulate a real node.js require for electron renderer modules that are reused in sandbox. Those are bundled in the executable and don't know they are running in a browserify environment.

tarruda replied:
What I mean is, electron modules such as crashReporter won't use this `require` function, so we need to tell browserify that when crashReporter requires `fs`, it is requiring `/lib/sandboxed_renderer/api/exports/fs.js` and not browserify shim.

miniak replied:
ok

tarruda commented on lib/sandboxed_renderer/init.js:
```diff
@@ -20,19 +21,26 @@ for (let prop of Object.keys(events.EventEmitter.prototype)) {
+  'path'
+])
+
+const loadedModules = new Map([
```
Technically, all modules are bundled. The ones that delegate to electron.remote are loaded lazily, but are still bundled.

tarruda commented:
@miniak With exception of taking preload path from webPreferences (which I want to do in a separate PR), I believe I've addressed all your comments. Let me know when you've finished your review so I can squash the fixup commits.

miniak commented on lib/sandboxed_renderer/init.js:
```diff
@@ -80,16 +92,15 @@ if (window.location.protocol === 'chrome-devtools:') {
   const preloadFn = geval(preloadWrapperSrc)
-  const {setImmediate} = require('timers')
-  preloadFn(preloadRequire, preloadProcess, Buffer, global, setImmediate)
+  const {setImmediate, clearImmediate} = require('timers')
```
👍

miniak commented on lib/sandboxed_renderer/init.js:
```diff
@@ -80,16 +92,15 @@ if (window.location.protocol === 'chrome-devtools:') {
-if (preloadPath) {
-  const preloadSrc = fs.readFileSync(preloadPath).toString()
-  const preloadWrapperSrc = `(function(require, process, Buffer, global, setImmediate) {
+if (preloadSrc) {
```
you need to check here if preloadSrc is an Error

miniak replied:
you are not logging the error here

tarruda replied:
I've used  `preloadError` for the error

miniak commented on lib/browser/rpc-server.js:
```diff
@@ -449,3 +450,21 @@ ipcMain.on('ELECTRON_BROWSER_WINDOW_CLOSE', function (event) {
+    try {
+      preloadSrc = fs.readFileSync(preloadPath).toString()
+    } catch (err) {
+      preloadSrc = err
```
maybe a separate variable? preloadError?

MarshallOfSound replied:
Also we need to be aware the `Error` objects to not JSON stringify very well by default, you'll need to manually pull out the `stack` and `message` attributes into a new object

MarshallOfSound commented on lib/sandboxed_renderer/init.js:
```diff
@@ -20,19 +21,26 @@ for (let prop of Object.keys(events.EventEmitter.prototype)) {
+  preloadSrc, webContentsId, platform, execPath, env
+} = electron.ipcRenderer.sendSync('ELECTRON_BROWSER_SANDBOX_LOAD', preloadPath)
+
+process.webContentsId = webContentsId
```
For safety

```js
Object.defineProperty(process, 'webContentsId', {
  configurable: false,
  value: webContentsId
})
```

This way nothing can ever override that value

tarruda commented on lib/sandboxed_renderer/init.js:
```diff
@@ -103,4 +106,6 @@ if (preloadSrc) {
   const {setImmediate, clearImmediate} = require('timers')
   preloadFn(preloadRequire, preloadProcess, Buffer, global, setImmediate, clearImmediate)
+} else if (preloadError) {
+  console.error(preloadError.stack)
```
@miniak is this how you wanted to handle preload error? I can't think of a better way right now, logging via `console.error` seems like a sane way to do it, since it would be available in the terminal if electron is started with `--enable-logging`

miniak commented on lib/browser/rpc-server.js:
```diff
@@ -449,3 +450,23 @@ ipcMain.on('ELECTRON_BROWSER_WINDOW_CLOSE', function (event) {
+    try {
+      preloadSrc = fs.readFileSync(preloadPath).toString()
+    } catch (err) {
+      preloadError = {stack: err.stack}
```
why don't you include name and message as well?

tarruda replied:
The `stack` property already contains the message and Error class name. Example from the repl:

```
> console.log(new Error('msg').stack)
Error: msg
    at repl:1:13
    at ContextifyScript.Script.runInThisContext (vm.js:50:33)
    at REPLServer.defaultEval (repl.js:240:29)
    at bound (domain.js:301:14)
    at REPLServer.runBound [as eval] (domain.js:314:12)
    at REPLServer.onLine (repl.js:468:10)
    at emitOne (events.js:121:20)
    at REPLServer.emit (events.js:211:7)
    at REPLServer.Interface._onLine (readline.js:282:10)
    at REPLServer.Interface._line (readline.js:631:8)
```

MarshallOfSound commented on lib/browser/rpc-server.js:
```diff
@@ -449,3 +450,23 @@ ipcMain.on('ELECTRON_BROWSER_WINDOW_CLOSE', function (event) {
+    try {
+      preloadSrc = fs.readFileSync(preloadPath).toString()
+    } catch (err) {
+      preloadError = {stack: err.stack}
```
I hate to be that person 😆 But welcome to javascript where you can `throw null`

This should check if we can get to `.stack` before actually trying or our error handling will error 😆

tarruda replied:
:+1:

tarruda commented:
Rebased/squashed commits

alexeykuzmin commented on lib/sandboxed_renderer/init.js:
```diff
@@ -20,19 +21,29 @@ for (let prop of Object.keys(events.EventEmitter.prototype)) {
+} = electron.ipcRenderer.sendSync('ELECTRON_BROWSER_SANDBOX_LOAD', preloadPath)
+
+Object.defineProperty(process, 'webContentsId', {
+  configurable: false,
```
`false` is a default value of the `configurable` attribute, no reasons to explicitly set it.
If you want to be explicit then you should also set `writable: false` (which is also a default value).

alexeykuzmin commented:
@tarruda Can you please backport the changes to `2-0-x`?

MarshallOfSound commented:
cc @tarruda ^^

tarruda commented:
:+1:
